### DISCUSSION TITLE
Error: Unexpected end of form at Multipart._final

### DISCUSSION THREAD

**Discussion by petewk** [expressjs/multer#1230] [Q&A]
Hi folks,

Am having the error above come up while trying to have Multer handle a file upload to my local storage, which I'll also later then push to S3 storage (once I get past the first hurdle).
I've wound up resorting to just using the same set up as shown in the simple example at https://www.npmjs.com/package/multer but am still having no luck.

Here's my express set up in case it's an issue with another package I'm depending on;

```
const express = require ('express');
const app = express();

const multer  = require('multer');
const upload = multer({ dest: 'uploads/' });

const bodyParser = require ('body-parser');
const path = require('path');
const fs = require('fs');
const port = process.env.PORT || 5000;
const fileupload = require('express-fileupload');
const AWS = require('aws-sdk');

const util  = require('util');

app.use(bodyParser.urlencoded({extended: true}));
app.use(bodyParser.json());

const s3 = new AWS.S3();
app.use(fileupload());

```

My React form component for selecting the files (the commented out authCode input is irrelevant for now): 

```
function FileUpload(){

    return(
        <>
            <h4>Upload files here</h4>
            <form action='/pdf' method='POST' enctype="multipart/form-data">
                {/* <input className="createFormInput" name="authCode" id="authCodeInput" placeholder='Enter auth code here' type="number" required></input> */}
                <input type='file' name="avatar"></input>
                <button>Submit</button>
            </form>
        </>
    )
}
```

and my express handling the post request

```
  app.post('/pdf', upload.single('avatar'), function (req, res, next) {

    if (err instanceof multer.MulterError){
      console.log("multer error")
    }
    // req.file is the `avatar` file
    // req.body will hold the text fields, if there were any
  })
```

Some pages seemed to suggest it was caused by the bodyParser, but it wasn't clear why or what was wrong with my set up. Any tips would be greatly appreciated.

Thanks

**Comment by Doc999tor** [comment]
1. The error comes from `buboy`, then it can't complete the `data` stream. I can't say if it's `multer` or `express-fileupload` throws the error, since both use `busboy` for file uploading. It looks like one of them is excessive - I bet you can deal all your file handling logic with `multer` 😊  
2. Text fields handling - you connect both `multer` (`busboy`) and `body-parser`, since both of them subcribe to `data.end` event, maybe `busboy` misses the end the of the stream

**Comment by agustinFeijoo** [comment]
Same error

**Comment by BAHIRABDULLA** [comment]
remove that fileUpload middleware in app.use . only multer need to upload file , that if you try to upload in s3 bucket .

**Comment by usamadevx** [comment]
I encountered an **Error**: Unexpected end of form originating from the **busboy** library, specifically within its **Multipart._final** method. While I'm using multer for file uploads, the issue isn't likely within multer itself, as multer integrates with (or uses) busboy internally to handle multipart form data. The problem often arises from conflicting Express middleware. **If you are also using the express-fileupload middleware globally (e.g., with app.use(fileUpload()) in your app.js or main server file)**, this is likely the cause. Both multer (via busboy) and express-fileupload attempt to parse the incoming form data stream. This conflict can lead to one middleware consuming the stream prematurely, causing busboy to signal an unexpected end. To resolve this, **_remove the global app.use(fileUpload()) middleware_** registration from your application setup.

**Comment by Uanela** [comment]
Same error when I try `multer().none()`

**Comment by Uanela** [comment]
The real problem is that probably we're trying to apply 2 middlewares to handle multipart/form-data input, and it was true for me, I using 2 multer.none() middlewares (just my own lib mistake was adding 2 instead of just one).

**Comment by bjohansebas** [comment]
Hey, I haven’t been able to reproduce the error. If someone can create a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), that would be great

**Comment by Tommy1223454** [comment]
One thing I’d double-check here is whether multiple middleware/packages are trying to handle the same multipart request/stream.

Once that part is stable, I’d also recommend doing trust decisions before the file is pushed further (local disk, S3, etc.), not after. In other words: parse -> inspect -> then store/promote.

I built a Node.js OSS package for that inspection layer:
https://github.com/pompelmi/pompelmi

It’s useful when you want more than “the upload completed” and need to screen untrusted files before they become part of your storage pipeline.

**Reply by Uanela** [reply]
The problem is exactly it once req.on("data", handler) there is no longer stream to read from.

**Comment by janjongboom** [comment]
Here's how we fixed it:

```diff
diff --git a/lib/make-middleware.js b/lib/make-middleware.js
--- a/lib/make-middleware.js
+++ b/lib/make-middleware.js
@@ -36,12 +36,15 @@ function makeMiddleware (setup) {
     var pendingWrites = new Counter()
     var uploadedFiles = []
 
+    function ignoreLateParserError () {}
+
     function done (err) {
       if (isDone) return
       isDone = true
       req.unpipe(busboy)
       process.nextTick(() => {
         busboy.removeAllListeners()
+        busboy.on('error', ignoreLateParserError)
       })
       next(err)
     }
```
